src/legacy/core_plugins/kibana/public/local_application_service/local_application_service.ts

```diff
@@ -0,0 +1,136 @@
+   *
+   * When an app is mounted, it is responsible for routing. The app
+   * won't be mounted again if the route changes within the prefix
+   * of the app (its id). It is fine to use whatever means for handling
+   * routing within the app.
+   *
+   * When switching to a URL outside of the current prefix, the app router
+   * shouldn't do anything because it doesn't own the routing anymore -
+   * the local application service takes over routing again,
+   * unmounts the current app and mounts the next app.
+   *
+   * @param app The app descriptor
+   */
+  register(app: App) {
+    this.apps.push(app);
+  }
+
+  /**
+   * Forwards every URL starting with `legacyAppId` to the same URL starting
+   * with `newAppId` - e.g. `/legacy/my/legacy/path?q=123` gets forwarded to
+   * `/newApp/my/legacy/path?q=123`.
+   *
+   * When setting the `keepPrefix` option, the new app id is simply prepended.
+   * The example above would become `/newApp/legacy/my/legacy/path?q=123`.
+   *
+   * This method can be used to provide backwards compatibility for URLs when
+   * renaming or nesting plugins. For route changes after the prefix, please
+   * use the routing mechanism of your app.
+   *
+   * @param legacyAppId The name of the old app to forward URLs from
+   * @param newAppId The name of the new app that handles the URLs now
+   * @param options Whether the prefix of the old app is kept to nest the legacy
+   * path into the new path
+   */
+  forwardApp(
+    legacyAppId: string,
+    newAppId: string,
+    options: { keepPrefix: boolean } = { keepPrefix: false }
+  ) {
+    this.forwards.push({ legacyAppId, newAppId, ...options });
+  }
+
+  /**
+   * Wires up listeners to handle mounting and unmounting of apps to
+   * the legacy angular route manager. Once all apps within the Kibana
+   * plugin are using the local route manager, this implementation can
+   * be switched to a more lightweight implementation.
+   *
+   * @param angularRouteManager The current `ui/routes` instance
+   */
+  apply(angularRouteManager: UIRoutes) {
+    this.apps.forEach(app => {
+      const wrapperElementId = this.idGenerator();
+      angularRouteManager.when(matchAllWithPrefix(app), {
+        outerAngularWrapperRoute: true,
+        reloadOnSearch: false,
+        reloadOnUrl: false,
+        template: `<div style="height:100%" id="${wrapperElementId}"></div>`,
+        controller($scope: IScope) {
+          const element = document.getElementById(wrapperElementId)!;
+          (async () => {
+            const onUnmount = await app.mount({ core: npStart.core }, { element, appBasePath: '' });
+            $scope.$on('$destroy', () => {
+              onUnmount();
+            });
+          })();
+        },
+      });
+    });
+
+    this.forwards.forEach(({ legacyAppId, newAppId, keepPrefix }) => {
+      angularRouteManager.when(matchAllWithPrefix(legacyAppId), {
+        redirectTo: (_params: unknown, path: string, search: string) => {
```

Comment:
A quick note while testing, when search is an object, this produces links like:
http://localhost:5601/tmz/app/kibana#/discover/doc/indexPattern/index?%5Bobject%20Object%5D&_g=()

Reply:
Ah, good catch, thanks. I will look into this.